rfcs/20230515-DynamicEmbedding.md

```diff
@@ -0,0 +1,251 @@
+lookups of data. 
+Variables with dynamic shapes are a type of data structure that allows
+for variables to have different shapes at different times. This can be
+useful for storing data that is constantly changing, such as the
+vocabulary of a language. Right now, with parameter server strategy
+variables cannot be placed on parameter servers. Mutable hash tables
+are always placed on the chief, which could have performance
+implications for lookups, inserts, and updates to the vocabulary.
+However, if we can add support for the TensorFlow distribute side
+to allow per-worker variable creation, this performance implication
+can be overcome.
+
+## Dependencies
+The proposed feature does not introduce any new dependencies. It is
+a stand-alone feature that can be used with any existing TensorFlow
+workflow. There is no need to modify any existing code or workflows
+to use this feature.
+
+## Engineering Impact
+This feature can add a small time overhead to update the dynamic
+vocabulary table, but this comes with improved performance of models
+and less user intervention to update vocabulary and restart training.
+Training can be continuous and with real-time data, and the model
+would continuously keep updating its vocabulary. This is beneficial
+because it allows the model to learn new input patterns, which can
+improve its accuracy and performance. Additionally, it reduces the
+amount of time and effort required to maintain the model, as the
+user does not need to manually update the vocabulary table or
+restart training every time new data is available. These benefits
+are particularly valuable in an online learning setting
+
+## Platforms and Environments
+    * GPU, TPU, CPU
+    * Asynchronous distributed training
+Synchronous distributed training
+
+## Best Practices
+The following are the best practices used so far:
+    * The users need to stop training the model and update the
+    vocabulary before restarting training.
+    * The vocabulary that needs to be provided to the model needs
+    to be generated by the user separately.
+The DynamicEmbedding layer is a new layer that enables users to
+train a model on a dataset with a dynamic vocabulary. This means
+that the vocabulary can change over time without the user having
+to stop training the model and update the vocabulary. The layer
+is simply used as any other Keras layer. The initial vocabulary
+can be provided or the layer will learn the whole vocabulary on
+its own.
+
+## Tutorials and Examples
+```
+from keras.layers import DynamicEmbedding  
+train_data = np.array([
+       ['a', 'j', 'c', 'd', 'e'],
+       ['a', 'h', 'i', 'j', 'b'],
+       ['i', 'h', 'c', 'j', 'e'],
+
+   ])
+   train_labels = np.array([0, 1, 2])
+   vocab = tf.constant(['a', 'b', 'c', 'd', 'e'])
+   eviction_policy = 'LFU'
+   # Define the model
+   model = keras.models.Sequential([
+       DynamicEmbedding(
+           input_dim=5,
+           output_dim=2,
+           input_length=5,
```

Comment:
Can we support the inputs with dynamic shapes?

Reply:
Excellent, thank you for your answer! I would like to know what the `input_dim` means. From my understanding, `input_dim` should be less or equal to the `vocabulary size`, which is fixed when training going on, is it right?

Reply:
input_dim should be vocabulary size

Reply:
> @rhdong, I would like to clarify that for the layer initialization `inp_dim` is vocabulary size(tried to keep it consistent with the [embedding layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding#:~:text=.Embedding(-,input_dim%2C,-output_dim%2C%0A%C2%A0%20%C2%A0%20embeddings_initializer) - . The input to the layer can be of any dynamic shape.

Hi @divyashreepathihalli, thank you for your clarification. I understand now, but I believe the `input_dim` setting would limit the features encoding space. In the dynamic embedding context, `input_dim` should be `std::numeric_limits<uint64_t>::max`.  I would try to explain it in one google doc.

Reply:
> @rhdong, I would like to clarify that for the layer initialization `inp_dim` is vocabulary size(tried to keep it consistent with the [embedding layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding#:~:text=.Embedding(-,input_dim%2C,-output_dim%2C%0A%C2%A0%20%C2%A0%20embeddings_initializer) - . The input to the layer can be of any dynamic shape.

I may think @divyashreepathihalli  may misunderstand the meaning of dynamic shape embedding. For example, there is a training feature input that is both large scale and sparse, such as USER_ID. If we apply the vocabulary method to USER_ID, it will only map USER_ID to the dimension of vocabulary size, which is a compression of the information dimension. Since the vocabulary size is fixed, this is still a static embedding. Dynamic embedding means that all inputs can be mapped non-conflicting through a hash function. The size of the dynamic embedding is not fixed because the USER_ID grows with the growth of the business.

Reply:
@rhdong  @MoFHeka thank you for the clarification. I tried to read up further. If I understand correctly you are looking for a dynamic vocabulary size and a dynamic embedding matrix as well, correct? One that would keep growing?

As of now our scope of work will be limited to maintaining a fixed size vocabulary and fixed embedding size, updating the vocabulary based on inputs received by the layer and eviction policies. The embedding values will be remapped whenever the vocabulary is updated based on input patterns (most frequently seen input, TTL, etc). If the input key is not in the vocab it will be mapped to a default value, however we keep track of these keys and add it to the vocab when the updates are done in the callback(new keys are added in the vocab by kicking out old keys based on eviction policies specified)

Reply:
@divyashreepathihalli It's my pleasure. Considering the practical scenario of dynamic embedding we reached out to, the hashtable-based dynamic vocabulary size would be a fundamental requirement. I guess one of the PROs of your current design is that there is no need to modify the `tf.optimizer`; that makes sense, but in addition to the considerations we discussed above, I'm also a little worried it will introduce the data consistency issue caused by decoupling the `embedding indexing` and `embedding looking up`, especially in eviction involved. As you know, adding atomic or lock mechanisms to ID and embedding is challenging when they are operated in two separate OPs.